File: 1. Harvard CS50/sentimental-credit/credit.py
# TODO
from cs50 import get_string


def main():
    # Prompt for input
    card = get_string("Number: ")

    # Calculate checksum
    checksum, length = calculate_card_checksum(card)

    # Check for card length and starting digits
    verifyCard(checksum, card, length)


def calculate_card_checksum(card_number):
    total_sum = 0
    length = 0
    current_digit = 0

    while current_digit != card_number:
        length += 1
        # Catch the last digit in the card
        current_digit = card_number[-length:]

        even_number = 0
        odd_sum = 0

        # Verify if the number is odd or even
        if length % 2:
            odd_sum = int(current_digit[0])
        else:
            first_sum = int(current_digit[0]) * 2
            # If the number is greater than 9, sum the digits
            if first_sum >= 10:
                digits = [int(digit) for digit in str(first_sum)]
                even_number += digits[0] + digits[1]
            else:
                even_number += first_sum

        total_sum += even_number + odd_sum

    return total_sum, length


def verifyCard(checksum, card, length):
    if 12 < length < 17 and checksum % 10 == 0:
        firstNumbers = int(card[:2])
        if (firstNumbers == 34 or firstNumbers == 37) and length == 15:
            print("AMEX")
        elif 50 < firstNumbers < 56 and length == 16:
            print("MASTERCARD")
        elif 40 <= firstNumbers < 50 and (length == 13 or length == 16):
            print("VISA")
        else:
            print("INVALID")
    else:
        print("INVALID")


main()

File: 1. Harvard CS50/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int calculateCard(long card);
void verifyCard(int checkSum, long card);

int main(void)
{
    // Prompt for input
    long card = get_long("Diga: ");

    // Calculate checksum
    int checkSum = calculateCard(card);

    // Check for card length and starting digits
    verifyCard(checkSum, card);
}

int calculateCard(long card)
{
    int length = 0;
    long totalSum = 0;
    long digit = 0;

    // Calculates the digits of the card
    for (long i = 10; digit != card; i *= 10)
    {
        // Add the length of the card
        length++;

        // Defines and calculates the digits
        digit = card % i;

        long firstSum = 0;
        long oddSum = 0;

        if (length % 2)
        {
            oddSum = digit / (i / 10);
        }
        else
        {
            long evenNumber = (digit / (i / 10)) * 2;

            if (evenNumber >= 10)
            {
                firstSum = firstSum + ((evenNumber / 10) + (evenNumber % 10));
            }
            else
            {
                firstSum = firstSum + evenNumber;
            }
        }

        // Calculates the final result
        totalSum = totalSum + (firstSum + oddSum);
        printf("%li\n", totalSum);
    }
    return totalSum;
}

void verifyCard(int checkSum, long card)
{
    int length = 0;
    long digit = 0;
    long i = 10;
    for (; digit != card; i *= 10)
    {
        length++;
        digit = card % i;
    }

    if (length > 12 && length < 17 && checkSum % 10 == 0)
    {
        long firstNumbers = digit / (i / 1000);
        if ((firstNumbers == 34 || firstNumbers == 37) && length == 15)
        {
            printf("AMEX\n");
        }
        else if (firstNumbers > 50 && firstNumbers < 56 && length == 16)
        {
            printf("MASTERCARD\n");
        }
        else if (firstNumbers >= 40 && firstNumbers < 50 && (length == 13 || length == 16))
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}
